passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/Protocol.kt

```diff
@@ -82,32 +81,33 @@ class Protocol @JvmOverloads constructor(
      */
     @Throws(IllegalArgumentException::class, ProtocolException::class, PheException::class)
     fun enrollAccount(password: String): CompletableFuture<EnrollResult> = GlobalScope.async {
-        if (password.isBlank()) Utils.shouldNotBeEmpty("password")
-
-        Passw0rdProtos.EnrollmentRequest.newBuilder().setVersion(currentVersion).build().run {
-            httpClient.firePost(
-                    this,
-                    HttpClientProtobuf.AvailableRequests.ENROLL,
-                    authToken = appToken,
-                    responseParser = Passw0rdProtos.EnrollmentResponse.parser()
-            ).let { response ->
-                val enrollResult = try {
-                    pheClients[response.version]!!.enrollAccount(response.response.toByteArray(),
-                                                                 password.toByteArray())
-                } catch (exception: PheException) {
-                    throw InvalidProofException()
-                }
-
-                val enrollmentRecord = Passw0rdProtos.DatabaseRecord
-                        .newBuilder()
-                        .setVersion(currentVersion)
-                        .setRecord(ByteString.copyFrom(enrollResult.enrollmentRecord))
-                        .build()
-                        .toByteArray()
-
-                EnrollResult(enrollmentRecord, enrollResult.accountKey)
-            }
+        requires(password.isNotBlank(), "password")
+
+        val enrollmentRequest = Passw0rdProtos.EnrollmentRequest.newBuilder().setVersion(currentVersion).build()
```

Comment:
Why don't you like kotlin's run and let functions? We can avoid mess with lots of local variables.

Reply:
Because:
1. It's enough hard to understand which value will return in the end (especially without IDE)
2. `run`, `let` and other kotlin's functions are often used in inappropriate way. It even required an [entire article](https://medium.com/@elye.project/mastering-kotlin-standard-functions-run-with-let-also-and-apply-9cd334b0ef84) to explain their usecases. In current case, I believe, code should look like
```kotlin
with (Passw0rdProtos.EnrollmentRequest.newBuilder().setVersion(currentVersion).build()) {
            run {
                val response = httpClient.firePost(
                        this,
                        HttpClientProtobuf.AvailableRequests.ENROLL,
                        authToken = appToken,
                        responseParser = Passw0rdProtos.EnrollmentResponse.parser()
                )
                val enrollResult = try {
                    pheClients[response.version]!!.enrollAccount(response.response.toByteArray(),
                            password.toByteArray())
                } catch (exception: PheException) {
                    throw InvalidProofException()
                }

                val enrollmentRecord = Passw0rdProtos.DatabaseRecord
                        .newBuilder()
                        .setVersion(currentVersion)
                        .setRecord(ByteString.copyFrom(enrollResult.enrollmentRecord))
                        .build()
                        .toByteArray()

                EnrollResult(enrollmentRecord, enrollResult.accountKey)
            }
        }
```
That's why I prefer simple linear code - it makes code more understandable.

Reply:
Agree, depending on that article it's better to use 'with' instead of 'run'. But we don't need inner 'run' in your example.
You can not understand which value will be returned without IDE even in simple case:
```
val value = someFunction()
```
I think that not using kotlin benefits to simplify code for...(who?) java developers? Kotlin got it's own nice ownings that make code more concise, with less local variables, etc.
As for me even without IDE code with standard kotlin functions looks understandable and pretty (sure, you have not to overuse it).
But, actually how many people will write a program without IDE nowadays?

Anyway, in this code I think you're probably right. I will take one more look at the your suggestion (:

Reply:
I mean Intellij Idea (2018.3.3 in my case) shows that result of `async` is `Passw0rdProtos...` and `EnrollResult...` is result of `let` extension work. While in fact `EnrollResult` is return value of `async`. It's quite difficult to understand at a glance :)

Reply:
Still it depends on your mindset (: 
I can just look at the last line of `let` block to see what is the result. As well if you got longer chain call - you'll look through the chains and get the last statement. If the return result is not in the end - it's just a bad design. I'm trying not to nest hundreds of function one inside of other. So it's not that hard to find out the result of 2-3 nested functions.
Yes, I got your position, and you're partly right. But as for me - it's nearly impossible to not to use IDE nowadays. So we can state that 90%+ developers will see this code through the IDE. Accordingly to this you can use both variants - yours and mine. And sure, for me - mine is closer and more understandable.
Nowadays it's not logical to suppose that somebody will work with your code using notepad. It's normally to suppose that your code will be used with some IDE.
I agree, that respectively to the article that you sent it's needed to change a few functions.
But while you `prefer simple linear code`, I prefer simpler code in my way (:

Reply:
Yeah I agree, of course, I'm not against your way :)

Reply:
Done :)